### ISSUE TITLE
More customisation possible?

### ISSUE THREAD

**Issue by zohozer** [cinatic/stocks-extension#49]
Hi. I do like a lot this software but will be great to have some more customisation options like:
- Custom fonts. Right now the fonts are barely visible in my screen, I need to get close to screen to be able to read the numbers.
- Option to show only the price without the (price change$ | percentage%)

**Comment by LAOzzy** [comment]
Another good addition would be some sort of notification when thresholds are met. Such as if a given stock passes a given price. Some rule-based notifications?

**Comment by zohozer** [comment]
> 

Indeed, an useful feature, but for me the font size is the most annoing. I need to stare too close to my monitor to be able to read the price of the stocks.

**Comment by jpnurmi** [comment]
Just hearing the word "template" results in an idea like
```
{name} {quote}{currency} ({change}{currency} | {percent}%)
```
but it doesn't translate well to multiple `St.Label` instances with different fonts, colors, alignment etc. Do the St widgets have something more suitable for this?

**Comment by cinatic** [comment]
that is what i have in my mind, each field would be a own `St.Label`, no color customization yet

I can imagine providing some modifiers like {change:small:bold:left} and a {linebreak} and then let's see how this will look like in the end. Actually i am not very confident with the positioning, y/x_align & expand often is not really doing what i expected

**Comment by jpnurmi** [comment]
How do you like this kind of variant that separates text formatting and template variables?

#
![image](https://user-images.githubusercontent.com/140617/166139252-c5114054-9511-410e-a852-9bbc1dc2db49.png)
```
<small><b>{name}</b><br/>{quote}{currency} ({change}{currency} | {percent}%)</small>
```

#
![image](https://user-images.githubusercontent.com/140617/166139254-41464ff3-b2ad-4665-91b6-3cea4a2709ba.png)
```
<p>{symbol}</p>{quote}<small>{currency} ({change}{currency} {percent}%)</small>
```

#
![image](https://user-images.githubusercontent.com/140617/166139257-8d0f19db-028d-4b33-a399-6a5cc01f1020.png)
```
<b>{symbol}<b> {quote}<small>{currency}</small>
```

#
![image](https://user-images.githubusercontent.com/140617/166139259-5591da60-60eb-4882-9c9b-ee4c3e718533.png)
```
{quote}
```
My personal favorite is a meaningless plain number to make it more pleasant to share the screen.

**Comment by cinatic** [comment]
Looks good, i looked around maybe it's even possible to put that straight into st label as markup

https://docs.gtk.org/gtk4/class.Label.html#markup-styled-text
https://gitlab.gnome.org/GNOME/mutter/-/issues/1324

**Comment by jpnurmi** [comment]
Nice. A quick attempt suggests that foreground colors are not respected but otherwise, it looks really promising.

**Comment by jpnurmi** [comment]
```
<span foreground="red">Red</span> <b>bold</b> <i>italic</i> <small>small</small> <sub>sub</sub> <sup>sup</sup>
```
![image](https://user-images.githubusercontent.com/140617/166158219-662517ee-2182-4785-b188-90c5bd63bbb5.png)

**Comment by jpnurmi** [comment]
I've removed stylesheets and trimmed the whole extension all the way down to:
```
function enable () {
  let button = new PanelMenu.Button()
  let label = new St.Label()
  label.get_clutter_text().set_markup('<span color="red">fg</span> <span background="red">bg</span> <b>bold</b>')
  button.add_actor(label)
  Main.panel.addToStatusArea('stocksMenu', button)
}
```
Still no success with the foreground color.
![Screenshot from 2022-05-05 19-47-14](https://user-images.githubusercontent.com/140617/166982901-32ab8959-0b96-4e05-be40-7b3278cea62f.png)
I'm only able to change the foreground color by setting a style class, but that applies to the whole label, of course... Do you have any ideas or should we go back to parsing the markup by hand?

I'm on Ubuntu 22.04 / GNOME 42. Could someone else give this a try?

**Comment by cinatic** [comment]
@jpnurmi 
i just tried it, muellner mentioned here https://gitlab.gnome.org/GNOME/mutter/-/issues/1324 that if you put a single whitespace the font color works as well

```
  label.get_clutter_text().set_markup(' <span color="red">fg</span> <span background="red">bg</span> <b>bold</b>')

```
![image](https://user-images.githubusercontent.com/2721787/166985665-f5a402d6-4d60-4347-8c55-40b7beacd092.png)

do you want to continue with this ticker? i think i will not be able to start until weekend or maybe even next weekend

**Comment by jpnurmi** [comment]
Great, that works! Sounds like we have a plan. No rush :)

**Comment by translucide** [comment]
I made a fork that might be of interest to you guys here, it does not allow font changes but it allows one to customize what is displayed in each ticker item such as quote, change$ and change%. Settings are per-stock in portfolios items, so each ticker item can be customized properly... 

Some more work is needed for fonts styles however.
